enhancements/olm/operator-must-gather.md

```diff
@@ -0,0 +1,137 @@
+5. Should be 100% functional when not in a openshift environment.
+
+### Non-Goals
+
+1. Anything related to the wider cluster. This is not meant to be a OCP debug tool.
+2. Specficics for a given operator. This is meant to help debug install of the operator, not the operator's functionality. 
+3. Handle low level install scenarios, such as manually creating a CSV after applying RBAC.
+
+## Proposal
+
+A library, in the [api](https://github.com/operator-framework/api/tree/master/pkg) repo that will be vendored by `OLM` and `operator-sdk`. This library will include all of the traversal logic for retrieving OLM Operators data. 
+
+The library will include a set of pflags that can be used by commands to create the options structure.
+
+OLM will create a must-gather image and productize it. Allowing customers to run this must gather when needed.
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### Constraints
+This enhancement is only meant for operators that are deployed by OLM. The entrypoint should be the `subscription`.
+
+### Risks and Mitigations
+- No environment variables will be outputed from a pod.
+- We will not save any secret data that may have been created. 
+
+## Design Details
+* Considering that many different CLI's will be interacting with the library, the library will create an interface that the CLI's will need to implement to write files to disk.
+* The library will have a default version of this interface internally
+
+```
+type ObjectWriter interface {
+  // object is the object to be written. The implmenter can choose the output format.
+  // file is the relative path, from the starting directory of the inspect command, for the object.
+  Write(object interface{}, file string) error
+
+  // Used to determine if the root directory to be used by the library is valid.
+  IsDirViable(dir string) error
+
+}
+```
+
+* The library options will allow for the implmenter to control what is searched and how
+
+```type OperatorInspectConfig struct {
+       // A kubeconfig to be used to contact the cluster
+       Kubeconfig *rest.Config
+       // PackageName used to filter to a particular operator by it's package name. If empty all packages will be used.
+       PackageName string
+       // Namespace to look for subscription objects
+       // if left blank, will search the entire cluster for subscription.
+       Namespace   string
+       Writer ObjectWriter
+}
+```
+
+* A Inspector interface will be returned
+```
+type Inspector interface {
+  Inspect() error
+}
+
+// Will return the inspector that can be used for insepcting operators from OLM.
+func New(configArgs ...InspectorArg) (Inspector, error)
+```
+
+The Inspector interface will be responsible for gather the following resource types:
+1. Subscription 
+2. InstallPlan
+3. Resources created from the install plan (CSV's, RBAC, CRD)
+4. Operator Deployment created by CSV
+
+
+### Test Plan
+- an e2e test will be used to validate that the command can:
```

Comment:
I think it will be easier to test if the fs is optional via a writer you pass in (tests can be written that just write to a buffer)